### PR TITLE
fixes #11723 - Build PXE Default fails

### PR DESCRIPTION
**Pull request by phirince**

### CODE REVIEW

**Comment by domcleal** [comment]
[test]

File: lib/foreman/renderer.rb [hunk 1]
```diff
@@ -54,7 +54,7 @@ def foreman_url(action = "built")
 
       url_for :only_path => false, :controller => "/unattended", :action => action,
               :protocol  => protocol, :host => host, :port => port, :script_name => path,
-              :token     => (@host.token.value unless @host.token.nil?)
+              :token     => (@host.token.value unless @host.try(:token).nil?)
```
Comment (unorthodoxgeek) [inline lib/foreman/renderer.rb:57]: don't understand this code. previous line has a local variable named host, this one has an instance variable... what's the difference?

Comment (phirince) [inline lib/foreman/renderer.rb:57]: First of all, the local variable is not part of my commit and is not relevant to this pull request.

If you look at a couple of lines above this line, you'll see the line:
host     = uri.host

so, the local  'host' variable is the host from which the template can be downloaded. Whereas the instance variable "@host" will tell you which host has requested the template. At least, that is my understanding.

-PP

**Comment by unorthodoxgeek** [comment]
seems to be ok for merge :+1:

**Comment by domcleal** [comment]
How did you manage to trigger this error?  Our default template can't do it, as far as I can tell.

Usually the default template uses the default_template_url helper and not foreman_url, which isn't even included in the safemode permitted helpers.

**Comment by phirince** [comment]
I was not aware of default_template_url helper. Is it documented somewhere? Apparently I missed it. So I had to disable safe mode and use foreman_url.

But still, I think this should be fixed. Even though it is not recommended, it _is_ possible to do from the UI and it does not look good on the software to show a cryptic error message to the user. Just my two cents.

-PP

**Comment by phirince** [comment]
I don't think default_template_url will be of much help to me. I wanted to set proxy.url to the foreman instance in PXELinux Global default. i.e. I want it to point to the proxy url and not the foreman master instance. I was trying to achieve that using foreman_url, but even that didn't help me. Is there any way to achieve this with the current version? If not I'll open a separate bug for that.

-PP

**Comment by domcleal** [comment]
> I was not aware of default_template_url helper. Is it documented somewhere? Apparently I missed it.

I don't think so, I just saw it's used in our default template as an alternative to foreman_url: https://github.com/theforeman/foreman/blob/develop/app/views/unattended/pxe/PXELinux_default.erb#L31, but it's specifically to generate URLs to host-independent templates, not much help for you now I see.

> So I had to disable safe mode and use foreman_url.

Right, I suspected as much!  I'm fine with fixing this now I understand how you're using it as the rest of the method does actually seem to be designed with this in mind.

Would you mind adding a unit test for this?  It ought to be quite simple, as you just run foreman_url without setting `@host`.  test/lib/foreman/renderer_test.rb might be the best place for it.

> I want it to point to the proxy url and not the foreman master instance. I was trying to achieve that using foreman_url, but even that didn't help me. Is there any way to achieve this with the current version? If not I'll open a separate bug for that.

http://projects.theforeman.org/issues/7828 tracks making foreman_url point to Foreman's base URL without a template URL etc.  There's no way to get the URL of any smart proxy, but http://projects.theforeman.org/issues/10286 tracked and started to implement this to match the proxy's template feature.

**Comment by phirince** [comment]
> http://projects.theforeman.org/issues/10286 tracked and started to implement this to match the proxy's template feature.

Actually the following two issues target the problem I'm having, IIRC:
http://projects.theforeman.org/issues/10286
http://projects.theforeman.org/issues/10244

Both of them are unassigned. Do you mind if I take ownership of them?

-PP

**Comment by domcleal** [comment]
> Both of them are unassigned. Do you mind if I take ownership of them?

Not at all!

Edit: I just increased your Redmine permissions, so if you had trouble, try again now.

**Comment by domcleal** [comment]
Merged as 536fd19, thanks @phirince.

**Pull request closed without merging**